File: develop-tools/src/main/java/com/lxp/tool/algorithm/OpenPGPHelper.java
package com.lxp.tool.algorithm;

import org.bouncycastle.bcpg.ArmoredOutputStream;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.openpgp.PGPCompressedData;
import org.bouncycastle.openpgp.PGPCompressedDataGenerator;
import org.bouncycastle.openpgp.PGPEncryptedData;
import org.bouncycastle.openpgp.PGPEncryptedDataGenerator;
import org.bouncycastle.openpgp.PGPEncryptedDataList;
import org.bouncycastle.openpgp.PGPException;
import org.bouncycastle.openpgp.PGPLiteralData;
import org.bouncycastle.openpgp.PGPLiteralDataGenerator;
import org.bouncycastle.openpgp.PGPObjectFactory;
import org.bouncycastle.openpgp.PGPOnePassSignature;
import org.bouncycastle.openpgp.PGPOnePassSignatureList;
import org.bouncycastle.openpgp.PGPPrivateKey;
import org.bouncycastle.openpgp.PGPPublicKey;
import org.bouncycastle.openpgp.PGPPublicKeyEncryptedData;
import org.bouncycastle.openpgp.PGPPublicKeyRing;
import org.bouncycastle.openpgp.PGPPublicKeyRingCollection;
import org.bouncycastle.openpgp.PGPSecretKey;
import org.bouncycastle.openpgp.PGPSecretKeyRing;
import org.bouncycastle.openpgp.PGPSecretKeyRingCollection;
import org.bouncycastle.openpgp.PGPSignature;
import org.bouncycastle.openpgp.PGPSignatureGenerator;
import org.bouncycastle.openpgp.PGPSignatureList;
import org.bouncycastle.openpgp.PGPSignatureSubpacketGenerator;
import org.bouncycastle.openpgp.PGPUtil;
import org.bouncycastle.openpgp.operator.PublicKeyDataDecryptorFactory;
import org.bouncycastle.openpgp.operator.jcajce.JcaKeyFingerprintCalculator;
import org.bouncycastle.openpgp.operator.jcajce.JcaPGPContentSignerBuilder;
import org.bouncycastle.openpgp.operator.jcajce.JcaPGPContentVerifierBuilderProvider;
import org.bouncycastle.openpgp.operator.jcajce.JcePBESecretKeyDecryptorBuilder;
import org.bouncycastle.openpgp.operator.jcajce.JcePGPDataEncryptorBuilder;
import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyDataDecryptorFactoryBuilder;
import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyKeyEncryptionMethodGenerator;
import org.bouncycastle.util.io.Streams;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.security.SecureRandom;
import java.security.Security;
import java.security.SignatureException;
import java.util.Date;
import java.util.Iterator;
import java.util.Objects;

public class OpenPGPHelper {
    private static final Logger LOGGER = LoggerFactory.getLogger(OpenPGPHelper.class);
    private static final BouncyCastleProvider BOUNCY_CASTLE_PROVIDER = new BouncyCastleProvider();

    private static OpenPGPHelper instance;

    private OpenPGPHelper() {
    }

    public static OpenPGPHelper getInstance() {

        if (instance == null) {
            instance = new OpenPGPHelper();
        }
        return instance;
    }


    public PGPPublicKey readPublicKey(InputStream in) throws IOException, PGPException {

        PGPPublicKeyRingCollection pgpPub = new PGPPublicKeyRingCollection(PGPUtil.getDecoderStream(in),
                new JcaKeyFingerprintCalculator());

        PGPPublicKey key = null;

        //
        // iterate through the key rings.
        //
        Iterator<PGPPublicKeyRing> rIt = pgpPub.getKeyRings();

        while (key == null && rIt.hasNext()) {
            PGPPublicKeyRing kRing = rIt.next();
            Iterator<PGPPublicKey> kIt = kRing.getPublicKeys();
            while (kIt.hasNext()) {  //(key == null && kIt.hasNext())
                PGPPublicKey k = kIt.next();
                LOGGER.debug("Is Encryption Key =" + k.isEncryptionKey());
                LOGGER.debug("Is Master Key =" + k.isMasterKey());
                LOGGER.debug("Is Revoked Key =" + k.hasRevocation());
                LOGGER.debug(" Key ID =" + k.getKeyID());
                LOGGER.debug(" Key Strength =" + k.getBitStrength());
                if (k.isEncryptionKey()) {
                    key = k;
                }
            }
        }

        if (key == null) {
            throw new IllegalArgumentException("Can't find encryption key in key ring.");
        }

        return key;
    }

    public PGPPrivateKey findSecretKey(PGPSecretKey pgpSecKey, char[] pass)
            throws PGPException {
        return pgpSecKey.extractPrivateKey(new JcePBESecretKeyDecryptorBuilder().setProvider(new BouncyCastleProvider()).build(pass));
    }

    static PGPPrivateKey findSecretKey(PGPSecretKeyRingCollection pgpSec, long keyID, char[] pass)
            throws PGPException {
        PGPSecretKey pgpSecKey = pgpSec.getSecretKey(keyID);

        if (pgpSecKey == null) {
            return null;
        }

        return pgpSecKey.extractPrivateKey(new JcePBESecretKeyDecryptorBuilder().setProvider(new BouncyCastleProvider()).build(pass));
    }

    public PGPSecretKey readSecretKey(InputStream input) throws IOException, PGPException {
        PGPSecretKeyRingCollection pgpSec = new PGPSecretKeyRingCollection(
                PGPUtil.getDecoderStream(input), new JcaKeyFingerprintCalculator());

        //
        // we just loop through the collection till we find a key suitable for encryption, in the real
        // world you would probably want to be a bit smarter about this.
        //

        Iterator<PGPSecretKeyRing> keyRingIter = pgpSec.getKeyRings();
        while (keyRingIter.hasNext()) {
            PGPSecretKeyRing keyRing = keyRingIter.next();

            Iterator<PGPSecretKey> keyIter = keyRing.getSecretKeys();
            while (keyIter.hasNext()) {
                PGPSecretKey key = keyIter.next();

                if (key.isSigningKey()) {
                    return key;
                }
            }
        }

        throw new IllegalArgumentException("Can't find signing key in key ring.");
    }

    public void encryptAndSign(OutputStream out, InputStream inputStream, PGPPublicKey encKey, PGPPrivateKey privateKey, String identify) {
        out = new ArmoredOutputStream(out);
        try {
            PGPEncryptedDataGenerator encGen =
                    new PGPEncryptedDataGenerator(
                            new JcePGPDataEncryptorBuilder(PGPEncryptedData.AES_256).setWithIntegrityPacket(true).setSecureRandom(
                                            new SecureRandom())
                                    .setProvider(new BouncyCastleProvider()));
            encGen.addMethod(new JcePublicKeyKeyEncryptionMethodGenerator(encKey).setProvider(new BouncyCastleProvider()));
            OutputStream encryptedOut = encGen.open(out, new byte[inputStream.available()]);

            PGPCompressedDataGenerator comData = new PGPCompressedDataGenerator(PGPCompressedData.ZIP);
            OutputStream compressedData = comData.open(encryptedOut);

            PGPSignatureGenerator sGen = new PGPSignatureGenerator(new JcaPGPContentSignerBuilder(
                    privateKey.getPublicKeyPacket().getAlgorithm(), PGPUtil.SHA512).setProvider(new BouncyCastleProvider()));
            sGen.init(PGPSignature.BINARY_DOCUMENT, privateKey);
            Iterator<String> it = encKey.getUserIDs();
            if (it.hasNext()) {
                PGPSignatureSubpacketGenerator spGen = new PGPSignatureSubpacketGenerator();
                spGen.addSignerUserID(false, it.next());
                sGen.setHashedSubpackets(spGen.generate());
            }
            //BCPGOutputStream bOut = new BCPGOutputStream(compressedData);
            sGen.generateOnePassVersion(false).encode(compressedData); // bOut

            PGPLiteralDataGenerator lGen = new PGPLiteralDataGenerator();
            OutputStream lOut = lGen.open(compressedData, PGPLiteralData.BINARY, identify, new Date(),
                    new byte[inputStream.available()]); //bOut

            byte[] data = inputStream.readAllBytes();
            lOut.write(data);
            sGen.update(data);

            lOut.close();
            lGen.close();

            sGen.generate().encode(compressedData);

            //bOut.close();

            comData.close();
            compressedData.close();

            encryptedOut.close();
            encGen.close();

            out.close();
        } catch (PGPException e) {
            if (e.getUnderlyingException() != null) {
                LOGGER.error(e.getUnderlyingException().getMessage(), e.getUnderlyingException());
            } else {
                LOGGER.error(e.getMessage(), e);
            }
        } catch (Exception e) {
            LOGGER.error(e.getMessage(), e);
        }
    }

    public void decryptAndVerify(InputStream in, OutputStream out, BufferedInputStream publicKeyStream, InputStream privateKeyStream, String password)
            throws Exception {
        Security.addProvider(new BouncyCastleProvider());
        PGPObjectFactory pgpF = new PGPObjectFactory(PGPUtil.getDecoderStream(in), new JcaKeyFingerprintCalculator());
        PGPEncryptedDataList enc;
        Object o = pgpF.nextObject();
        //
        // the first object might be a PGP marker packet.
        //
        if (o instanceof PGPEncryptedDataList) {
            enc = (PGPEncryptedDataList) o;
        } else {
            enc = (PGPEncryptedDataList) pgpF.nextObject();
        }
        //
        // find the secret key
        //
        Iterator<?> it = enc.getEncryptedDataObjects();
        PGPPublicKeyEncryptedData pbe = null;
        PGPSecretKeyRingCollection pgpSec = new PGPSecretKeyRingCollection(
                PGPUtil.getDecoderStream(privateKeyStream), new JcaKeyFingerprintCalculator());
        PGPPrivateKey sKey = null;

        while (sKey == null && it.hasNext()) {
            pbe = (PGPPublicKeyEncryptedData) it.next();
            sKey = findSecretKey(pgpSec, pbe.getKeyID(), password.toCharArray());
        }
        if (Objects.isNull(sKey)) {
            throw new IllegalArgumentException("fail to find secret key");
        }
        PublicKeyDataDecryptorFactory b = new JcePublicKeyDataDecryptorFactoryBuilder()
                .setProvider(new BouncyCastleProvider()).setContentProvider(new BouncyCastleProvider()).build(sKey);

        InputStream clear = pbe.getDataStream(b);

        PGPObjectFactory plainFact = new PGPObjectFactory(clear, new JcaKeyFingerprintCalculator());

        Object message = plainFact.nextObject();

        PGPOnePassSignatureList onePassSignatureList = null;
        PGPSignatureList signatureList = null;
        PGPCompressedData compressedData;
        ByteArrayOutputStream actualOutput = new ByteArrayOutputStream();
        while (message != null) {
            if (message instanceof PGPCompressedData) {
                compressedData = (PGPCompressedData) message;
                plainFact = new PGPObjectFactory(compressedData.getDataStream(), new JcaKeyFingerprintCalculator());
                message = plainFact.nextObject();
            }

            if (message instanceof PGPLiteralData) {
                // have to read it and keep it somewhere.
                Streams.pipeAll(((PGPLiteralData) message).getInputStream(), actualOutput);
            } else if (message instanceof PGPOnePassSignatureList) {
                onePassSignatureList = (PGPOnePassSignatureList) message;
            } else if (message instanceof PGPSignatureList) {
                signatureList = (PGPSignatureList) message;
            } else {
                throw new PGPException("message unknown message type.");
            }
            message = plainFact.nextObject();
        }
        actualOutput.close();
        byte[] output = actualOutput.toByteArray();
        PGPPublicKey publicKeyIn = null;
        if (onePassSignatureList == null || signatureList == null) {
            throw new PGPException("Poor PGP. Signatures not found.");
        } else {
            for (int i = 0; i < onePassSignatureList.size(); i++) {
                PGPOnePassSignature ops = onePassSignatureList.get(0);
                PGPPublicKeyRingCollection pgpPub = new PGPPublicKeyRingCollection(PGPUtil.getDecoderStream(publicKeyStream), new JcaKeyFingerprintCalculator());
                PGPSignature signature = signatureList.get(i);
                publicKeyIn = pgpPub.getPublicKey(signature.getKeyID());
                if (publicKeyIn != null) {
                    ops.init(new JcaPGPContentVerifierBuilderProvider().setProvider(BOUNCY_CASTLE_PROVIDER), publicKeyIn);
                    ops.update(output);
                    if (ops.verify(signature)) {
                        Iterator<?> userIds = publicKeyIn.getUserIDs();
                        while (userIds.hasNext()) {
                            String userId = (String) userIds.next();
                            LOGGER.debug(userId);
                        }
                    } else {
                        throw new SignatureException("Signature verification failed");
                    }
                }
            }
        }

        if (pbe.isIntegrityProtected() && !pbe.verify()) {
            throw new PGPException("Data is integrity protected but integrity is lost.");
        } else if (publicKeyIn == null) {
            throw new SignatureException("Signature not found");
        } else {
            out.write(output);
            out.flush();
            out.close();
        }
    }
}

File: STUDY-CSHARP/Utility/OpenPGPHelper.cs
using Org.BouncyCastle.Bcpg;
using Org.BouncyCastle.Bcpg.OpenPgp;
using Org.BouncyCastle.Security;
using Org.BouncyCastle.Utilities.IO;

namespace Utility
{
    public class OpenPGPHelper
    {
        private static NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();
        private static OpenPGPHelper? instance;
        private OpenPGPHelper()
        {

        }

        public static OpenPGPHelper GetInstance()
        {
            if (instance == null)
            {
                instance = new OpenPGPHelper();
            }
            return instance;
        }

        public PgpPublicKey ReadPublickey(Stream inputStream)
        {
            PgpPublicKeyRingBundle pubRings = new PgpPublicKeyRingBundle(PgpUtilities.GetDecoderStream(inputStream));
            PgpPublicKey key = null;

            foreach (PgpPublicKeyRing kRing in pubRings.GetKeyRings())
            {
                foreach (PgpPublicKey k in kRing.GetPublicKeys())
                {
                    logger.Debug("Is Encryption Key =" + k.IsEncryptionKey);
                    logger.Debug("Is Master Key =" + k.IsMasterKey);
                    logger.Debug("Is Revoked Key =" + k.IsRevoked());
                    logger.Debug("Key ID =" + k.KeyId);
                    logger.Debug("Key Strength =" + k.BitStrength);
                    if (k.IsEncryptionKey)
                    {
                        key = k;
                        break;
                    }
                }
                if (key != null)
                {
                    break;
                }
            }
            if (key == null)
            {
                throw new System.Exception("Can't find encryption key in key ring.");
            }

            return key;
        }

        public PgpPrivateKey FindSecretKey(PgpSecretKey pgpSecretKey, char[] pass)
        {
            return pgpSecretKey.ExtractPrivateKey(pass);
        }

        public static PgpPrivateKey FindSecretKey(PgpSecretKeyRingBundle pgpSec, long keyID, char[] pass)
        {
            PgpSecretKey pgpSecKey = pgpSec.GetSecretKey(keyID);

            if (pgpSecKey == null)
            {
                return null;
            }

            return pgpSecKey.ExtractPrivateKey(pass);
        }

        public PgpSecretKey ReadSecretKey(Stream inputStream)
        {
            PgpSecretKeyRingBundle pgpSec = new PgpSecretKeyRingBundle(PgpUtilities.GetDecoderStream(inputStream));

            // To Be Improved
            // we just loop through the collection till we find a key suitable for encryption, in the real
            // world you would probably want to be a bit smarter about this.

            foreach (var keyRing in pgpSec.GetKeyRings())
            {
                foreach (var key in keyRing.GetSecretKeys())
                {
                    if (key.IsSigningKey)
                    {
                        return key;
                    }
                }
            }
            throw new System.Exception("Can't find signing key in key ring.");
        }

        public void EncryptAndSign(Stream outputStream, Stream inputStream, PgpPublicKey encryptionKey, PgpPrivateKey signingKey, string identify)
        {
            outputStream = new ArmoredOutputStream(outputStream);
            try
            {
                PgpEncryptedDataGenerator encryptedDataGenerator = new PgpEncryptedDataGenerator(SymmetricKeyAlgorithmTag.Aes256, true, new SecureRandom());
                encryptedDataGenerator.AddMethod(encryptionKey);

                Stream encryptedOut = encryptedDataGenerator.Open(outputStream, new byte[1 << 16]);

                PgpCompressedDataGenerator compressedDataGenerator = new PgpCompressedDataGenerator(CompressionAlgorithmTag.Zip);
                Stream compressedOut = compressedDataGenerator.Open(encryptedOut);

                PgpSignatureGenerator signatureGenerator = new PgpSignatureGenerator(signingKey.PublicKeyPacket.Algorithm, HashAlgorithmTag.Sha512);
                signatureGenerator.InitSign(PgpSignature.BinaryDocument, signingKey);
                foreach (string userId in encryptionKey.GetUserIds())
                {
                    PgpSignatureSubpacketGenerator subpacketGenerator = new PgpSignatureSubpacketGenerator();
                    subpacketGenerator.SetSignerUserId(false, userId);
                    signatureGenerator.SetHashedSubpackets(subpacketGenerator.Generate());
                    break;
                }
                signatureGenerator.GenerateOnePassVersion(false).Encode(compressedOut);

                PgpLiteralDataGenerator literalDataGenerator = new PgpLiteralDataGenerator();
                Stream literalOut = literalDataGenerator.Open(compressedOut, PgpLiteralData.Binary, identify, DateTime.Now, new byte[1 << 16]);

                byte[] buffer = new byte[1 << 16];
                int bytesRead;
                while ((bytesRead = inputStream.Read(buffer, 0, buffer.Length)) > 0)
                {
                    literalOut.Write(buffer, 0, bytesRead);
                    signatureGenerator.Update(buffer, 0, bytesRead);
                }

                literalOut.Close();
                literalDataGenerator.Close();

                signatureGenerator.Generate().Encode(compressedOut);

                compressedOut.Close();
                compressedDataGenerator.Close();

                encryptedOut.Close();
                encryptedDataGenerator.Close();

                outputStream.Close();
            }
            catch (PgpException e)
            {
                if (e.InnerException != null)
                {
                    logger.Error(e.InnerException.Message, e.InnerException);
                }
                else
                {
                    logger.Error(e.Message, e);
                }
            }
            catch (Exception e)
            {
                logger.Error(e.Message, e);
            }
        }

        public void DecryptAndVerify(Stream inputStream, Stream outputStream, Stream publicKeyStream, Stream privateKeyStream, string password)
        {
            PgpObjectFactory pgpF = new PgpObjectFactory(PgpUtilities.GetDecoderStream(inputStream));
            // PgpEncryptedDataList enc = pgpF.NextPgpObject() as PgpEncryptedDataList;

            PgpEncryptedDataList enc;
            Object o = pgpF.NextPgpObject();
            if(o.GetType() == typeof(PgpEncryptedDataList))
            {
                enc = (PgpEncryptedDataList) o;
            }
            else
            {
                enc = (PgpEncryptedDataList) pgpF.NextPgpObject();
            }

            PgpPublicKeyEncryptedData pbe = null;
            PgpSecretKeyRingBundle pgpSec = new PgpSecretKeyRingBundle(PgpUtilities.GetDecoderStream(privateKeyStream));
            PgpPrivateKey sKey = null;
            foreach (PgpPublicKeyEncryptedData data in enc.GetEncryptedDataObjects())
            {
                pbe = data;
                sKey = FindSecretKey(pgpSec, pbe.KeyId, password.ToCharArray());
            }
            if(sKey == null)
            {
                throw new System.Exception("fail to find secret key");
            }

            Stream clear = pbe.GetDataStream(sKey);
            PgpObjectFactory plainFact = new PgpObjectFactory(clear);
            object message = plainFact.NextPgpObject();
            PgpOnePassSignatureList onePassSignatureList = null;
            PgpSignatureList signatureList = null;
            PgpCompressedData compressedData;
            MemoryStream actualOutput = new MemoryStream();

            while (message != null)
            {
                if (message is PgpCompressedData)
                {
                    compressedData = (PgpCompressedData)message;
                    plainFact = new PgpObjectFactory(compressedData.GetDataStream());
                    message = plainFact.NextPgpObject();
                }

                if (message is PgpLiteralData)
                {
                    // have to read it and keep it somewhere.
                    Streams.PipeAll(((PgpLiteralData)message).GetInputStream(), actualOutput);
                }
                else if (message is PgpOnePassSignatureList)
                {
                    onePassSignatureList = (PgpOnePassSignatureList)message;
                }
                else if (message is PgpSignatureList)
                {
                    signatureList = (PgpSignatureList)message;
                }
                else
                {
                    throw new PgpException("message unknown message type.");
                }

                message = plainFact.NextPgpObject();
            }

            actualOutput.Close();
            byte[] output = actualOutput.ToArray();
            PgpPublicKey publicKeyIn = null;

            if (onePassSignatureList == null || signatureList == null)
            {
                throw new PgpException("Poor PGP. Signatures not found.");
            }
            else
            {
                PgpPublicKeyRingBundle pgpPub = new PgpPublicKeyRingBundle(PgpUtilities.GetDecoderStream(publicKeyStream));
                for (int i = 0; i < onePassSignatureList.Count; i++)
                {
                    PgpOnePassSignature ops = onePassSignatureList[0];
                    PgpSignature signature = signatureList[i];
                    publicKeyIn = pgpPub.GetPublicKey(signature.KeyId);

                    if (publicKeyIn != null)
                    {
                        ops.InitVerify(publicKeyIn);
                        ops.Update(output);
                        if (ops.Verify(signature))
                        {
                            logger.Debug(string.Join(",", publicKeyIn.GetUserIds()));
                        }
                        else
                        {
                            throw new SignatureException("Signature verification failed");
                        }
                    }
                }
            }

            if (pbe.IsIntegrityProtected() && !pbe.Verify())
            {
                throw new PgpException("Data is integrity protected but integrity is lost.");
            }
            else if (publicKeyIn == null)
            {
                throw new SignatureException("Signature not found");
            }
            else
            {
                outputStream.Write(output, 0, output.Length);
            }
        }
    }
}